cpu/stm32_common/periph/i2c_1.c

```diff
@@ -0,0 +1,497 @@
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            /* end transmission */
+            _stop(i2c);
+            return -ETIMEDOUT;
+        }
+    }
+
+    DEBUG("[i2c] read_regs: send start sequence\n");
+
+    /* send start sequence and slave address */
+    ret = _start(i2c, address, 1, 0, flags);
+    if (ret < 0) {
+        return ret;
+    }
+
+    tick = TICK_TIMEOUT;
+    /* wait for ack */
+    while (!(i2c->ISR & I2C_ISR_TXIS) && tick--) {
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            /* end transmission */
+            _stop(i2c);
+            return -ETIMEDOUT;
+        }
+    }
+
+    DEBUG("[i2c] read_regs: Write register to read\n");
+    i2c->TXDR = reg;
+
+    /* send repeated start sequence, read registers and end transmission */
+    DEBUG("[i2c] read_regs: ACK received, send repeated start sequence\n");
+    return i2c_read_bytes(dev, address, data, length, 0);
+}
+
+int i2c_write_bytes(i2c_t dev, uint16_t address, const void *data,
+                    size_t length, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+
+    int ret = 0;
+
+    if (!(flags & I2C_NOSTART)) {
+        DEBUG("[i2c] write_bytes: start transmission\n");
+        /* start transmission and send slave address */
+        ret = _start(i2c, address, length, I2C_FLAG_WRITE, flags);
+        if (ret < 0) {
+            return ret;
+        }
+    }
+
+    DEBUG("[i2c] write_bytes: write the data\n");
+    /* send out data bytes */
+    ret = _write(i2c, data, length);
+    if (ret < 0) {
+        DEBUG("[i2c] write_bytes: nothing was written\n");
+        return ret;
+    }
+
+    if (!(flags & I2C_NOSTOP)) {
+        DEBUG("[i2c] write_bytes: end transmission\n");
+        /* end transmission */
+        ret = _stop(i2c);
+        if (ret < 0) {
+            return ret;
+        }
+    }
+
+    return ret;
+}
+
+int i2c_write_regs(i2c_t dev, uint16_t address, uint16_t reg, const void *data,
+                   size_t length, uint8_t flags)
+{
+    assert(dev < I2C_NUMOF);
+
+    uint16_t tick = TICK_TIMEOUT;
+    I2C_TypeDef *i2c = i2c_config[dev].dev;
+
+    int ret = 0;
+
+    if ((flags & I2C_REG16) || (flags & I2C_ADDR10)) {
+        return -EOPNOTSUPP;
+    }
+
+    /* Check to see if the bus is busy */
+    while ((i2c->ISR & I2C_ISR_BUSY) && tick--) {
+        if ((i2c->ISR & ERROR_FLAG) || !tick) {
+            return -ETIMEDOUT;
+        }
+    }
+
+    /* start transmission and send slave address */
+    /* increase length because our data is register+data */
+    ret = _start(i2c, address, length + 1, I2C_FLAG_WRITE, flags);
+    if (ret < 0) {
+        return ret;
+    }
+
+    /* send register number */
+    DEBUG("[i2c] write_regs: ACK received, write reg into DR\n");
+    i2c->TXDR = reg;
+
+    /* write out data bytes */
+    ret = _write(i2c, data, length);
+    if (ret < 0) {
+        return ret;
+    }
+
+    /* end transmission */
+    ret = _stop(i2c);
```

Comment:
No support for I2C_NOSTOP here ?

Reply:
You're right but it may have some use case with those weird I2C devices in the wild.